File: Contest_Yandex_School/2.cpp
#include <bits/stdc++.h>
using namespace std;

// Рекурсивный парсер: синтаксис grammar:
// Expression := signedTerm {( '+' | '-' ) signedTerm}
// signedTerm := {('+' | '-')} factor
// factor := number | '(' Expression ')'
long long parseExpression(const string &s, int &i);
long long parseSignedTerm(const string &s, int &i);
long long parseFactor(const string &s, int &i);

long long parseExpression(const string &s, int &i) {
    long long val = parseSignedTerm(s, i);
    int n = s.size();
    while (true) {
        // пропускаем пробелы
        while (i < n && s[i] == ' ') i++;
        if (i >= n || s[i] == ')') break;
        char op = s[i];
        if (op != '+' && op != '-') break;
        i++;
        long long term = parseSignedTerm(s, i);
        if (op == '+') val += term;
        else           val -= term;
    }
    return val;
}

long long parseSignedTerm(const string &s, int &i) {
    int n = s.size();
    while (i < n && s[i] == ' ') i++;
    long long sign = 1;
    // Обрабатываем несколько подряд стоящих знаков
    while (i < n && (s[i] == '+' || s[i] == '-')) {
        if (s[i] == '-') sign = -sign;
        i++;
        while (i < n && s[i] == ' ') i++;
    }
    long long factorVal = parseFactor(s, i);
    return sign * factorVal;
}

long long parseFactor(const string &s, int &i) {
    int n = s.size();
    while (i < n && s[i] == ' ') i++;
    if (i < n && s[i] == '(') {
        i++;  // пропустить '('
        long long val = parseExpression(s, i);
        while (i < n && s[i] == ' ') i++;
        if (i < n && s[i] == ')') i++;
        return val;
    } else {
        // Число
        long long num = 0;
        while (i < n && s[i] == ' ') i++;
        while (i < n && isdigit(s[i])) {
            num = num * 10 + (s[i] - '0');
            i++;
        }
        return num;
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    string expr;
    if (!getline(cin, expr)) return 0;
    int idx = 0;
    long long result = parseExpression(expr, idx);
    cout << result;
    return 0;
}

File: Contest_Yandex_School/2.py
def evaluate(s: str) -> int:
    s = s.replace(' ', '')
    stack = []
    result = 0
    op = '+'
    i = 0
    n = len(s)
    
    while i < n:
        if s[i] in '+-':
            op = s[i]
            i += 1
        else:
            # Parse sign
            sign = 1
            j = i
            while j < n and s[j] in '+-':
                if s[j] == '-':
                    sign *= -1
                j += 1
            i = j
            
            # Parse term: number or '('
            if i < n and s[i] == '(':
                stack.append((result, op))
                result = 0
                op = '+'
                i += 1
            else:
                # Parse number
                num = 0
                while i < n and s[i].isdigit():
                    num = num * 10 + int(s[i])
                    i += 1
                num *= sign
                
                # Apply current operator
                if op == '+':
                    result += num
                else:
                    result -= num
        
        # Check for closing parenthesis
        while i < n and s[i] == ')':
            prev_res, prev_op = stack.pop()
            if prev_op == '+':
                prev_res += result
            else:
                prev_res -= result
            result = prev_res
            i += 1
    
    return result